File: knotlink-sdk-cpp/tests/test_signal_subscribe_integration.cpp
/*
 * KnotLink SDK - C++ Integration Test
 * Signal-Subscribe pattern: Subscriber + Sender
 */

#include "../knotlink/SignalSubscriber.hpp"
#include "../knotlink/SignalSender.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <atomic>
using namespace knotlink;

static std::atomic<bool> shutdown{false};

void run_subscriber() {
    SignalSubscriber sub("app.knotlinksdk.test", "test_signal");
    sub.setOnDataReceivedCallback([](const std::string& data) {
        std::cout << "[Subscriber] Received: " << data << std::endl;
    });
    std::cout << "[Subscriber] Listening..." << std::endl;
    while (!shutdown) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void run_sender() {
    std::this_thread::sleep_for(std::chrono::seconds(1));
    SignalSender sender;
    sender.setConfig("app.knotlinksdk.test", "test_signal");
    sender.emitt("Hello from Sender!");
    std::cout << "[Sender] Signal sent." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(2));
    shutdown = true;
}

int main() {
    std::thread t(run_subscriber);
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    run_sender();
    t.join();
    return 0;
}

File: knotlink-sdk-python/tests/integration/test_signal_subscribe_integration.py
#!/usr/bin/env python3
# test_signal_subscribe_integration.py
import time
import threading
from knotlink import SignalSubscriber, SignalSender

_shutdown = threading.Event()

def run_subscriber():
    def on_received(data):
        print(f"[Subscriber] Received: {data}")
    sub = SignalSubscriber("app.knotlinksdk.test", "test_signal")
    sub.set_RecvFunc(on_received)
    print("[Subscriber] Listening...")
    try:
        while not _shutdown.wait(timeout=1):
            pass
    finally:
        sub.disconnect()
        print("[Subscriber] Disconnected.")

def run_sender():
    _shutdown.wait(timeout=1)  # 等待 subscriber 准备就绪
    sender = SignalSender()
    sender.set_config("app.knotlinksdk.test", "test_signal")
    try:
        sender.emitt("Hello from Sender!")
        print("[Sender] Signal sent.")
    finally:
        sender.disconnect()
        print("[Sender] Disconnected.")
        _shutdown.set()  # 通知 subscriber 退出

if __name__ == "__main__":
    t = threading.Thread(target=run_subscriber, daemon=True)
    t.start()
    run_sender()
    t.join(timeout=5)